src/data/finsupp/equiv_dfinsupp.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import data.dfinsupp
+import data.finsupp.basic
+import algebra.group.pi
+import group_theory.submonoid.operations
+
+/-!
+# Equivalence between `finsupp` and `dfinsupp` constrained to the range of `single i`
+
+## Main definitions
+
+* `finsupp.single_submonoid` - the submonoid corresponding to the range of `finsupp.single`.
+* `finsupp.single_submonoid.of` - the obvious map into `finsupp.single_submonoid`, a `→+`.
+* `finsupp.equiv_dfinsupp_single_submonoid` - the `add_equiv` between a `finsupp` and a `dfinsupp`
+  where each fibre is `finsupp.single_submonoid`.
+
+## Implementation notes
+
+Lean has a very hard time finding some typeclasses in this file - so in many cases we have to
+explicitly build them.
+
+-/
+
+variables {ι : Type*} {M : Type*}
+
+noncomputable theory
+
+namespace finsupp
+
+variables (M)
+
+/-- The submonoid corresponding to the range of `finsupp.single`. -/
+abbreviation single_submonoid [add_monoid M] (i : ι) : add_submonoid (ι →₀ M) :=
```

Comment:
Does it make sense to call this `single_mrange`?

Reply:
If we do that we end up with a name conflict for `linear_map.range` and `subgroup.range` when I inevitably extend this to subgroups and submodules - both are after the coveted `single_range` name, but one will have to be prefixed. I suppose that's not too big a deal, but it's why I went for the easy option of just putting the type at the end.

Reply:
I understand, but how about `single_grange` and `single_lrange` for those others? I think it's useful if the name mentions `range`. Now it sounds like you are putting a `submonoid` instance on something called `single`, or so...